DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx

```diff
@@ -233,18 +262,103 @@ struct TagTwoProngDisplacedVertices {
+    } else if (channel == aod::tagandprobe::TagChannels::DsOrDplusToKKPi) {
+      pdgDecayMothers.push_back(constants::physics::Pdg::kDPlus);
+      pdgDecayMothers.push_back(constants::physics::Pdg::kDS);
+    } else if (channel == aod::tagandprobe::TagChannels::DstarPlusToDzeroPi || channel == aod::tagandprobe::TagChannels::DstarMinusToDzeroBarPi) {
+      pdgDecayMothers.push_back(constants::physics::Pdg::kDStar);
+    }
+  }
+
+  /// Check if the given tag tracks belong to a D meson
+  /// \param firstTrack candidate
+  /// \param SecondTrack candidate
+  /// \param particlesMc McParticles table
+  /// \param channel decay channel
+  /// \return 1 if the TagParticles belong to the chosen D meson
+  ///                 AND the D meson has the correct probe as one of its daughters;
+  ///                 0 al the other cases.
+  template <typename P, typename T>
+  int getTagOrigin(T const& firsTrack,
+                   T const& secondTrack,
+                   P const& particlesMc,
+                   const uint8_t channel,
+                   std::vector<int>& pdgDecayMothers)
+  {
+    auto pdgTagMother = 0;
+    int pdgProbeParticle = -1;
+
+    if (channel == aod::tagandprobe::TagChannels::DplusToKPiPi) {
+      pdgTagMother = constants::physics::Pdg::kDPlus;
+      pdgProbeParticle = 321; // Ka
+    } else if (channel == aod::tagandprobe::TagChannels::DsOrDplusToKKPi) {
+      pdgTagMother = constants::physics::Pdg::kPhi;
+      pdgProbeParticle = 211; // Pi
+    } else {
+      pdgTagMother = constants::physics::Pdg::kD0;
+      pdgProbeParticle = 211; // Pi
+    }
+
+    if (!firsTrack.has_mcParticle() || !secondTrack.has_mcParticle()) {
+      return -999;
+    } else {
+      auto firstMcTrack = firsTrack.template mcParticle_as<aod::McParticles>();
+      auto secondMcTrack = secondTrack.template mcParticle_as<aod::McParticles>();
+      auto firstTrackMotherId = RecoDecay::getMother(particlesMc, firstMcTrack, pdgTagMother, true);
+      auto secondTrackMotherId = RecoDecay::getMother(particlesMc, secondMcTrack, pdgTagMother, true);
+
+      if ((firstTrackMotherId == secondTrackMotherId) && (firstTrackMotherId != -1)) {
+        auto particleMother = particlesMc.rawIteratorAt(firstTrackMotherId);
+
+        /// π±π± for D± → K∓π±π± decays
+        if (channel == aod::tagandprobe::TagChannels::DplusToKPiPi) {
+          auto particleMother = particlesMc.rawIteratorAt(firstTrackMotherId);
+          auto daughters = particleMother.template daughters_as<aod::McParticles>();
+          // Check if the probe is within the mother's particle daughters
+          for (auto& daughter : daughters) {
+            if (std::abs(daughter.pdgCode()) == pdgProbeParticle) {
+              return 1;
+            }
+          }
+        }
+
+        ///  K∓K± for φ from Ds± or D± → φπ± decays
+        ///  K∓π± for D0 from D±* → D0π± decays
+        else {
+          for (auto pdgGrandMother : pdgDecayMothers) {
```

Comment:
in case you use `DsOrDplusToKKPi` you return 1 either if the kaon has a `Ds` or a `D+` as mother, but in this way is not distinguishable. Is it what you want?

Reply:
Thank you for the comment! 
For the time being we are following the Fabrizio Grosa recipe where he does not  make any selection on the mother for this particular decay channel. Here we are just looking that both Kaons come from the Phi.
https://indico.cern.ch/event/1371921/contributions/5773861/attachments/2787104/4859629/AOTTrack_Dmeson_TagAndProbe.pdf